File: test/scripts/11-lock-contention/correct.py
#!/usr/bin/env python3
"""
CORRECT: Lock-free with local accumulation, rare lock.
Resume: "Prefer Lock-Free" / reduce contention
"""
import os
import signal
import threading
import time
NUM_THREADS = 32
ITERATIONS = 50_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
lock = threading.Lock()
counter = 0
stop = False

def increment():
    global counter
    local = 0
    while not stop:
        for _ in range(ITERATIONS):
            local += 1
        with lock:
            counter += local
        local = 0

def main():
    global stop
    print("CORRECT: Local accum, lock once per batch")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    threads = [threading.Thread(target=increment) for _ in range(NUM_THREADS)]
    for t in threads:
        t.start()
    time.sleep(DURATION_SEC)
    stop = True
    for t in threads:
        t.join(timeout=2)
    print(f"Stopped. counter={counter}")

if __name__ == "__main__":
    main()

File: test/scripts/11-lock-contention/correct.js
/**
 * CORRECT: Atomics.add - lock-free atomic increment
 */
const { Worker, isMainThread, workerData } = require('worker_threads');
const NUM_WORKERS = 8;
const ITERATIONS = 10_000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);

if (isMainThread) {
  const sab = new SharedArrayBuffer(8);
  const stopSab = new SharedArrayBuffer(4);
  const arr = new Int32Array(sab);
  console.log('CORRECT: Atomics.add (lock-free)');
  const workers = [];
  for (let i = 0; i < NUM_WORKERS; i++) {
    workers.push(new Promise((r) => {
      const w = new Worker(__filename, { workerData: { sab, stopSab } });
      w.on('exit', r);
    }));
  }
  setTimeout(() => {
    Atomics.store(new Int32Array(stopSab), 0, 1);
    Promise.all(workers).then(() => console.log('Stopped. counter=', arr[0]));
  }, DURATION_SEC * 1000);
  process.on('SIGINT', () => Atomics.store(new Int32Array(stopSab), 0, 1));
} else {
  const { sab, stopSab } = workerData;
  const arr = new Int32Array(sab);
  const stop = new Int32Array(stopSab);
  while (!Atomics.load(stop, 0)) {
    for (let i = 0; i < ITERATIONS; i++) {
      Atomics.add(arr, 0, 1);
    }
  }
}
